Index: ../alsa-lib-1.0.24/src/pcm/pcm_direct.c
===================================================================
--- ../alsa-lib-1.0.24/src/pcm/pcm_direct.c	(revision 10340)
+++ ../alsa-lib-1.0.24/src/pcm/pcm_direct.c	(working copy)
@@ -424,13 +424,21 @@
 		close(dmix->server_fd);
 		return ret;
 	}
-	
+
+#ifdef __uClinux__
+	ret = vfork();
+#else
 	ret = fork();
+#endif
 	if (ret < 0) {
 		close(dmix->server_fd);
 		return ret;
 	} else if (ret == 0) {
+#ifdef __uClinux__
+		ret = vfork();
+#else
 		ret = fork();
+#endif
 		if (ret == 0)
 			server_job(dmix);
 		_exit(EXIT_SUCCESS);
