--- libpcap-1.0.0/Makefile.in
+++ libpcap-1.0.0/Makefile.in
@@ -399,10 +399,7 @@
 #
 pcap-config: pcap-config.in Makefile
 	@rm -f $@ $@.tmp
-	sed -e 's|@includedir[@]|$(includedir)|g' \
-	    -e 's|@libdir[@]|$(libdir)|g' \
-	    -e 's|@DEPLIBS[@]|$(DEPLIBS)|g' \
-	    $(srcdir)/pcap-config.in >$@.tmp
+	./config.status --file=$@.tmp:$<
 	mv $@.tmp $@
 	chmod a+x $@
 
--- libpcap-1.0.0/pcap-config.in
+++ libpcap-1.0.0/pcap-config.in
@@ -4,13 +4,19 @@
 # Script to give the appropriate compiler flags and linker flags
 # to use when building code that uses libpcap.
 #
+
+prefix="@prefix@"
+exec_prefix="@exec_prefix@"
+includedir="@includedir@"
+libdir="@libdir@"
+
 case "$1" in
 
 --cflags)
-	echo "-I @includedir@"
+	echo "-I$includedir"
 	;;
 
 --libs)
-	echo "-L @libdir@ -lpcap @DEPLIBS@"
+	echo "-L$libdir -lpcap @DEPLIBS@"
 	;;
 esac
