Index: btsco-0.5/alsa-plugins/headsetd/headsetd.c
===================================================================
--- btsco-0.5/alsa-plugins/headsetd/headsetd.c	(revision 9358)
+++ btsco-0.5/alsa-plugins/headsetd/headsetd.c	(working copy)
@@ -36,6 +36,13 @@
 	printf("\theadsetd [-n not_daemon]\n");
 }
 
+#ifdef __uClinux__
+static void _daemon(int nochdir, int noclose)
+{
+	daemon(nochdir, noclose);
+}
+#endif
+
 int main(int argc, char **argv)
 {
 	int opt;
@@ -55,6 +62,9 @@
 	}
 
 	if (daemonize) {
+#ifdef __uClinux__
+		_daemon(0, 0);
+#else
 		int ret = fork();
 		if (ret > 0) {
 			exit(0);
@@ -74,6 +84,7 @@
 	
 			chdir("/");
 		}
+#endif
 	}
 
 	umask(0077);
